plugin/hotswap-agent-spring-plugin/src/main/java/org/hotswap/agent/plugin/spring/reload/SpringChangedAgent.java

```diff
@@ -0,0 +1,197 @@
+            result |= springChangedAgent.addClass(clazz);
+        }
+        return result;
+    }
+
+    public static boolean addChangedXml(URL xmlUrl) {
+        for (SpringChangedAgent springChangedAgent : springChangeAgents.values()) {
+            springChangedAgent.addXml(xmlUrl);
+        }
+        return true;
+    }
+
+    public static boolean addChangedProperty(URL property) {
+        for (SpringChangedAgent springChangedAgent : springChangeAgents.values()) {
+            springChangedAgent.addProperty(property);
+        }
+        return true;
+    }
+
+    private boolean isIgnorePackage(Class<?> clazz) {
+        String className = clazz.getName();
+        for (String ignorePackage : IGNORE_PACKAGES) {
+            if (className.startsWith(ignorePackage)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static void reload(long changeTimeStamps) {
+        int reloadCount = waitingReloadCount.incrementAndGet();
+        // avoid reload too much times, allow 2 tasks into waiting queue
+        if (reloadCount > 2) {
+            LOGGER.trace("Spring reload is already scheduled, skip this time:{}", changeTimeStamps);
+            waitingReloadCount.decrementAndGet();
+            return;
+        }
+        try {
+            for (SpringChangedAgent springChangedAgent : springChangeAgents.values()) {
+                // ensure reload only once, there is one lock.
+                springChangedAgent.reloadAll(changeTimeStamps);
+            }
+        } finally {
+            waitingReloadCount.decrementAndGet();
+        }
+    }
+
+    boolean addClass(Class clazz) {
+        if (clazz == null || isIgnorePackage(clazz)) {
+            return false;
+        }
+        springReload.addClass(clazz);
+        return true;
+    }
+
+    void addProperty(URL property) {
+        springReload.addProperty(property);
+    }
+
+    void addXml(URL xml) {
+        springReload.addXml(xml);
+    }
+
+    void addNewBean(BeanDefinitionRegistry registry, BeanDefinitionHolder beanDefinitionHolder) {
+        springReload.addScanNewBean(registry, beanDefinitionHolder);
+    }
+
+    private void reloadAll(long changeTimeStamps) {
+        try {
+            doReload(changeTimeStamps);
+        } catch (InterruptedException e) {
+            LOGGER.warning("reload spring failed: {}", e, ObjectUtils.identityToString(defaultListableBeanFactory));
+        }
+    }
+
+    private void doReload(long changeTimeStamps) throws InterruptedException {
+        boolean isLockAcquired = reloadLock.tryLock(1, TimeUnit.SECONDS);
+        if (isLockAcquired) {
+            try {
+                LOGGER.trace("Spring reload: {} at timestamps '{}'", ObjectUtils.identityToString(defaultListableBeanFactory), changeTimeStamps);
+                springReload.reload(changeTimeStamps);
+            } finally {
+                reloadLock.unlock();
+            }
+        } else {
```

Comment:
Not sure why this can happen. We should guarentee a SpringChangedAgent instance will not run doReload again before the running doReload finishes.